examples/caustics.dx

```diff
@@ -0,0 +1,475 @@
+'Caustic Rendering
+
+Based on https://benedikt-bitterli.me/tantalum/
+
+'## Rendering utilities
+
+AbsPoint = (Int & Int)
+RelPoint = (Float & Float)
+
+def pixel (w:Int)?-> (h:Int)?-> (x:Int) (y:Int) : (Fin w & Fin h) = (x@_, y@_)
+
+def drawIntoCanvas (height:Int) (width:Int)
+                   (f: ref:Type ?-> Ref ref ((Fin width & Fin height) =>out)
+                                ->{Accum ref} Unit)
+                   : Fin height=>Fin width=>out =
+  canvas = snd $ withAccum f
+  for y x. canvas.(x, y)
+
+def checkerboard (height:Int)?-> (width:Int)?-> (ref:Type)?->
+      (canvas:Ref ref ((Fin width & Fin height)=>Float))
+      :{Accum ref} Unit =
+  h_over_two = height `idiv` 2
+  -- note: this works but can't be reduced so it has to be a def (?)
+  for i:(Fin width) j:(Fin h_over_two).
+    canvas!(pixel (ordinal i) ((2 * ordinal j) + (ordinal i `rem` 2))) += 1.0
+  ()
+
+:plotmat drawIntoCanvas 20 20 checkerboard
+
+square = \x:Float. x * x
+
+-- Rasterize a segment of a line in a single logical pixel centered between
+-- four physical pixels.
+def rasterLineSegmentInFourPixels (height:Int)?-> (width:Int)?-> (ref:Type)?->
+      (_: VSpace out)?=>
+      (pixel_x: Int) -- 0 <= pixel_x < width - 1
+      (pixel_y: Int) -- 0 <= pixel_y < height - 1
```

Comment:
Oh, interesting. They are actually `Fin (width-1)`. Would it make sense to represent the canvas size as `Unit|Fin width_minus_1` in this case?

Reply:
Hm. Maybe? But then we would have to change all the other functions (`drawIntoCanvas`, `rasterLineSegment`, etc) to accept non-Fin-based tables. I'm not sure if that's a net improvement or not. What do you think?

(also, technically, this function works fine if you give it things out of bounds, it just does nothing in that case. So it's not strictly speaking an invariant.)

Reply:
Yeah... I'm just a little concerned with the number of index<->int casts in this code. I would consider that an usual case for Dex code, and it will likely cause us to deoptimize a bunch of stuff. I'd rather remove as many of them as we can.